WebAssembly/Instructions/Int64ShiftLeft.cs

```diff
@@ -1,25 +1,35 @@
+using System.Reflection.Emit;
+using WebAssembly.Runtime.Compilation;
+
 namespace WebAssembly.Instructions
 {
     /// <summary>
     /// Sign-agnostic shift left.
     /// </summary>
-    public class Int64ShiftLeft : ValueTwoToOneInstruction
+    public class Int64ShiftLeft : SimpleInstruction
     {
         /// <summary>
         /// Always <see cref="OpCode.Int64ShiftLeft"/>.
         /// </summary>
         public sealed override OpCode OpCode => OpCode.Int64ShiftLeft;
 
-        private protected sealed override WebAssemblyValueType ValueType => WebAssemblyValueType.Int64;
-
-        private protected sealed override System.Reflection.Emit.OpCode EmittedOpCode =>
-            System.Reflection.Emit.OpCodes.Shl;
-
         /// <summary>
         /// Creates a new  <see cref="Int64ShiftLeft"/> instance.
         /// </summary>
         public Int64ShiftLeft()
         {
         }
+
+        internal sealed override void Compile(CompilationContext context)
+        {
+            context.PopStackNoReturn(this.OpCode, WebAssemblyValueType.Int64, WebAssemblyValueType.Int64);
+
+            //Unlike WASM, CIL OpCodes.Shl requires the shift amount to be int32 or native int
+            //See: https://docs.microsoft.com/en-us/dotnet/api/system.reflection.emit.opcodes.shl?view=net-5.0
+            context.Emit(OpCodes.Conv_I4);  //Convert shift amount into int32
```

Comment:
64-bit shifts are supported on 64-bit .NET I believe.  Some kind of check here to remove this opcode when the platform supports it would be ideal...

Reply:
I added this probably because Mono (Unity) did not support 64-bit shift even on 64-bit.
Your idea of checking sounds nice.
However, considering Mono, `Environment.Is64BitProcess` is not enough. We have to actually run a 64-bit shift and check whether `InvalidProgramException` occurs.

Unfortunately, I am busy on my project, so I cannot quickly implement it...

Reply:
According to the spec you've linked, `native int` is a valid shift type, so you could just cast to `context.Emit(OpCodes.Conv_I);` and that would do the right thing on the platform automatically.

Reply:
If 64-bit mono's conv_i converts it to 64 bit, doesn't that put us in the same situation?

Reply:
@RyanLamansky Depends on whether the `InvalidProgramException` is being triggered by the stack type being `Int64` instead of the two allowed types of `Int32` and `native int` or whether mono just plainly doesn't deal with a shift amount larger than 32bit.

Reply:
From results of an experiment ( https://gist.github.com/tana/465af214d7cf463bbc364c4048cf1d0a ), the `native int` solution successfully worked on Mono (.NET Core accepted 64-bit shift amount, but Mono required conversion into `int32` or `native int`).
So I'm going to change `Conv_I4` of `i64` shift instructions to `Conv_I`.